fpdf/line_break.py

```diff
@@ -0,0 +1,237 @@
+from collections import namedtuple
+
+SOFT_HYPHEN = "\u00ad"
+HYPHEN = "\u002d"
+SPACE = " "
+NEWLINE = "\n"
+
+
+class Fragment:
+    def __init__(self, style, underlined, characters=None):
+        self.characters = [] if characters is None else characters
+        self.style = style
+        self.underline = underlined
+
+    @classmethod
+    def from_string(cls, string, style, underlined):
+        return cls(style, underlined, list(string))
+
+    def add_character(self, character):
+        self.characters.append(character)
+
+    def trim(self, index):
+        self.characters = self.characters[:index]
+
+    @property
+    def string(self):
+        return "".join(self.characters)
+
+    def __eq__(self, other):
+        return (
+            self.characters == other.characters
+            and self.style == other.style
+            and self.underline == other.underline
+        )
+
+
+text_line = namedtuple(
+    "text_line",
+    ("fragments", "text_width", "number_of_spaces_between_words", "justify"),
+)
+
+space_hint = namedtuple(
+    "break_hint",
+    (
+        "original_fragment_index",
+        "original_character_index",
+        "current_line_fragment_index",
+        "current_line_character_index",
+        "width",
+        "number_of_spaces",
+    ),
+)
+
+hyphen_hint = namedtuple(
+    "hyphen_hint",
+    space_hint._fields
+    + (
+        "character_to_append",
+        "character_to_append_width",
+        "character_to_append_style",
+        "character_to_append_underline",
+    ),
+)
+
+
+class CurrentLine:
+    def __init__(self):
+        self.fragments = []
+        self.width = 0
+        self.number_of_spaces = 0
+
+        # automatic break hints
+        self.space_break_hint = None
+        self.hyphen_break_hint = None
+
+    def add_character(
+        self,
+        character,
+        character_width,
+        style,
+        underline,
+        original_fragment_index,
+        original_character_index,
+    ):
+        assert character != NEWLINE
```

Comment:
I'm fine with defensive programming, but out of curiosity: why this check?

Reply:
when user of this class tries to append `NEWLINE` character to the object of this class -> something is definitely wrong.
User is expected to create a new object of this class, and fill it with readable characters

Reply:
Thanks for the explanation!